src/coreclr/vm/interoplibinterface_objc.cpp

```diff
@@ -78,31 +184,99 @@ BOOL QCALLTYPE ObjCBridgeNative::TrySetGlobalMessageSendCallback(
 
 namespace
 {
-    void* objc_alloc(void* self, void* sel)
+    bool TryGetReferenceTrackingScratchMemory(_In_ OBJECTREF object, _Out_ void** scratch)
     {
-        return NULL;
+        CONTRACTL
+        {
+            NOTHROW;
+            GC_NOTRIGGER;
+            PRECONDITION(CheckPointer(scratch));
+        }
+        CONTRACTL_END;
+
+        SyncBlock* syncBlock = object->PassiveGetSyncBlock();
+        if (syncBlock == NULL)
+            return false;
+
+        InteropSyncBlockInfo* interopInfo = syncBlock->GetInteropInfoNoCreate();
+        if (interopInfo == NULL)
+            return false;
+
+        // If no scratch memory is allocated, then the instance is not
+        // being tracked.
+        void* scratchLocal = interopInfo->GetReferenceTrackingScratchMemory();
+        if (scratchLocal == NULL)
+            return false;
+
+        *scratch = scratchLocal;
+        return true;
     }
+}
 
-    void objc_dealloc(void* self, void* sel)
+bool ObjCBridgeNative::IsTrackedReference(_In_ OBJECTREF object, _Out_ bool* isReferenced)
+{
+    CONTRACTL
     {
+        NOTHROW;
+        GC_NOTRIGGER;
+        PRECONDITION(CheckPointer(isReferenced));
+    }
+    CONTRACTL_END;
+
+    *isReferenced = false;
 
+    void* scratchMemory;
+    if (!TryGetReferenceTrackingScratchMemory(object, &scratchMemory))
+        return false;
+
+    _ASSERTE(g_IsReferencedCallback != NULL);
+    int result = g_IsReferencedCallback(scratchMemory);
+
+    *isReferenced = (result != 0);
+    return true;
+}
+
+void ObjCBridgeNative::OnBackgroundGCStarted()
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
     }
+    CONTRACTL_END;
+
+    if (g_BeginEndCallback != NULL)
```

Comment:
Note that xamarin-macios will take a global lock inside this callback: https://github.com/xamarin/xamarin-macios/blob/daa7e123e212eb3c9e0f30d951b66c0fe254d754/runtime/runtime.m#L337 . 

Taking this global lock while the background GC is in progress will frequently turn the background GC into blocking GC. The execution will tend to get blocked on this global lock that will be only released once the background GC is finished.

This may be as a shortcut to just hack some prototype together, but I do not think it acceptable as a shipping design.

Reply:
I am not very familiar with the purpose - is this to block prevent something while GC in progress or to do something while GC is in progress? I think it might be the second, since we ignore ephemeral GCs. 

Then, perhaps this begin/end should be around the blocking stage of background GC (when we do dependent handles and such) , or around entire GC when it is a blocking gen2.

Reply:
This is one of the reasons: https://github.com/xamarin/xamarin-macios/blob/daa7e123e212eb3c9e0f30d951b66c0fe254d754/runtime/runtime.m#L1941-L2006

Reply:
@rolfbjarne Yep. That one i have been able to understand - the comments are very helpful. The one I am more confused about is https://github.com/xamarin/xamarin-macios/blob/daa7e123e212eb3c9e0f30d951b66c0fe254d754/runtime/trampolines.m#L112-L114. This one, https://github.com/xamarin/xamarin-macios/blob/daa7e123e212eb3c9e0f30d951b66c0fe254d754/runtime/trampolines.m#L103-L106, seems similar but doesn't have the lock so the use case isn't obvious to me.

Reply:
If anybody can think of another solution than the lock, I'm happy to try it out.

Reply:
@AaronRobinsonMSFT it's only an issue if we call `release` on the native handle, those two cases call `retain` (because `release` can cause the object to be deallocated, which becomes a problem if the gc callback wants to call `retainCount` - but that's not a problem after `retain`).

Reply:
>but that's not a problem after retain

@rolfbjarne Okay. So I get everything you said above. But if that is true, then what is the purpose of surrounding the call with the lock? It would seem that locking isn't going to ensure that `i` isn't deallocated here. 

```ObjectiveC
xamarin_framework_peer_lock ();
[i retain];
xamarin_framework_peer_unlock ();
```

I started by looking for all instances where the "peer lock" was utilized. There seem to be two instances. The one below and the one in the managed ref release function. The one above is confusing me because it is a `retain` and not a `release`.

Reply:
Spoke offline. The above scenario is described as follows:

> Conceptually, it's these actions:
> 
> 1. Create managed wrapper object for a native object. 
> 2. Create a handle to this managed object in native code in some form (for Mono we use a MonoObject* pointer, but it can be any kind of representation), where this handle is the only thing keeping the GC from collecting the managed object.
> 3. Call retain on the native object.
> 4. Release/destroy the handle from 2.
> 5. The toggleref callback checks the retainCount of the native object.
> 6. The GC collects everything it can. 
> 
> The following order is the race condition:
> 
> 1-2-5-3-4-6
> 
> 1. Create managed wrapper object for a native object. 
> 2. Create a handle to this managed object in native code in some form (for Mono we use a MonoObject* pointer on the stack, but it can be any kind of representation), where this handle is the only thing keeping the GC from collecting the managed object.
> 5. The toggleref callback checks the retainCount of the native object (getting 1 and telling the GC that native code does not have a reference to the managed object).
> 3. Call retain on the native object.
> 4. Release/destroy the handle from 2.
> 6. The GC collects everything it can (which will collect the managed object, because the toggle ref callback said "don't need it", and there's nothing else keeping it alive)
> 
>

Reply:
> Create a handle to this managed object in native code in some form (for Mono we use a MonoObject* pointer

Is it important that the handle is created in native code - is this different from just keeping the managed object reference in managed code? There should be no difference for CoreCLR. There may be a difference for Mono due to how conservative GC stack scanning and embedding APIs work.

Reply:
No, it doesn't matter exactly where the handle is created, but I think that's missing the point.

The problem is the synchronization between releasing that handle, and code in the toggle ref callback. For Mono the handle is implicitly released by returning from the stack frame (this line of code makes sure the handle is considered alive up until that point: https://github.com/xamarin/xamarin-macios/blob/daa7e123e212eb3c9e0f30d951b66c0fe254d754/runtime/trampolines.m#L118 - other types of handles might need explicit release at that point).

Let's say we use a GCHandle (where the managed function returned a GCHandle of the managed object that was created), in which case this code block:

https://github.com/xamarin/xamarin-macios/blob/daa7e123e212eb3c9e0f30d951b66c0fe254d754/runtime/trampolines.m#L108-L119

could look something like this (`retval_gchandle` is the GCHandle the managed function returned):

```c
id i = xamarin_get_handle (retval_gchandle, exception_gchandle);
if (*exception_gchandle != 0) {
	mono_gchandle_free (retval_gchandle);
	return NULL;
}

xamarin_framework_peer_lock ();
[i retain];
xamarin_framework_peer_unlock ();
if (!retain)
	[i autorelease];

mono_gchandle_free (retval_gchandle);
return i;
```

Without the lock, the following scenario is possible:

a) Thread A is just about to execute [i retain].
b) Thread B is running the GC, and calls the toggle ref callback.
c) Thread B: The toggle ref callback sees that the retainCount of `i` is 1, and tells the GC that the object can be collected.
d) Thread A calls`[i retain]`, to prevent the object from being collected.
e) Thread B: The GC collects the object.